### PR TITLE
Support .editorconfig

### LINKED ISSUE #8534: EditorConfig
[EditorConfig](https://editorconfig.org) is somewhat of an industry-standard these days. Although not a priority for me, you might want to take into consideration being compatible with existing EditorConfig configurations.

### PR DESCRIPTION
**Pull request by SomeoneToIgnore**
Closes https://github.com/zed-industries/zed/issues/8534
Supersedes https://github.com/zed-industries/zed/pull/16349

Potential concerns:
* we do not follow up to the `/` when looking for `.editorconfig`, only up to the worktree root. 
Seems fine for most of the cases, and the rest should be solved generically later, as the same issue exists for settings.json
* `fn language` in `AllLanguageSettings` is very hot, called very frequently during rendering. We accumulate and parse all `.editorconfig` file contents beforehand, but have to go over globs and match these against the path given + merge the properties still.
This does not seem to be very bad, but needs more testing and potentially some extra caching. 

Release Notes:

- Added .editorconfig support

### CODE REVIEW

**Comment by BuonOmo** [comment]
Yeaaaaaaaaaah 🎉 

Thank you 🌮

File: crates/language/src/language_settings.rs [hunk 1]
```diff
@@ -810,13 +819,27 @@ impl InlayHintSettings {
...
+        language_name: Option<&LanguageName>,
+        cx: &'a AppContext,
+    ) -> Cow<'a, LanguageSettings> {
+        let settings = language_name
+            .and_then(|name| self.languages.get(name))
+            .unwrap_or(&self.defaults);
+
+        let editorconfig_properties = location.and_then(|location| {
+            cx.global::<SettingsStore>()
+                .editorconfg_properties(location.worktree_id, location.path)
```
[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply crates/language/src/language_settings.rs:834]: Thanks for spotting.

Both the local and multiplayer tests check the correctness of this, and some manual testing + profiling was made by me on repos with nested .editorconfig files.

**Comment by cweagans** [comment]
This is one of the only feature requests that I've been watching closely, so I'm unclear on the release process. Now that this is merged, how does this make its way to the version of Zed running on my machine? I see that it's in 0.159.0-pre, but I'm not sure what the lead time is from that point to an actual stable release that gets pulled down through the auto-update functionality.

**Comment by notpeter** [comment]
This is in Zed 0.159.x branch which is the current Zed Preview which released in the last hour. You can download it from here: https://zed.dev/releases/preview and try it out in Zed Preview today.  (You can safely run Zed Preview side by side with Zed Stable).

Next Wednesday (2024-10-30) 0.159.x will become the next Zed Stable release.

More documentation here: https://zed.dev/docs/development/releases

**Comment by cweagans** [comment]
Nice! Thank you so much for the information. Really appreciate it!

**Comment by lode** [comment]
I just saw https://github.com/zed-industries/zed/releases/tag/v0.159.0-pre. I was expecting to see the editorconfig at the general improvements instead of at language. I'm not sure whether this is fully automated, but I think it would be good to move this to the general section for the real release. Maybe this is my bias :) but I think this deserves bigger visibility for people waiting for this.

**Comment by BuonOmo** [comment]
:+1: also because this is not a language improvement (e.g: we do not start supporting `.editorconfig` syntax or language) but a setting configuration, the _languages_ section makes it harder to understand that.

**Comment by maxiim3** [comment]
Nice thank you !

**Comment by SomeoneToIgnore** [comment]
`.editorconfig` always overrides whatever Zed settings, so, yes?

**Comment by mibanez** [comment]
That was what I thought, thanks, still it makes me noise why to override a setting that is not  file format related but just a visual style preference.

**Comment by BuonOmo** [comment]
@SomeoneToIgnore I think this actually might be a problem for multiple users: `.editorconfig` is here to enforce a writing standard but not a way to use your ide. Maybe we could consider not setting soft wrap but wrap guides instead ? That would hint a user to respect editorconfig without changing their habits? @mibanez wdyt?

Also I don't know if we're transmitting information from zed settings to language formatters, but if that is the case, then that is where the line length is interesting (no soft wrap, but hard wrap made by formatter directly!)

**Comment by SomeoneToIgnore** [comment]
Seems simple to fix, if that's what's desired, please remove this block and check if it's working as expected for you: 

https://github.com/zed-industries/zed/blob/d90770c673bd19afcd88c2b35ccf4e711f02a18d/crates/language/src/language_settings.rs#L900-L904

**Comment by mibanez** [comment]
> Also I don't know if we're transmitting information from zed settings to language formatters, but if that is the case, then that is where the line length is interesting (no soft wrap, but hard wrap made by formatter directly!)

That would be awesome. We are using python with Ruff's formatter, its [line-length](https://docs.astral.sh/ruff/settings/#line-length) rule is not a hard upper bound and a few exceptions are allowed some times. In any case, the wrap guide sounds nice to me.

> please remove this block and check if it's working as expected for you

I'm trying this as soon as I can, ty!

**Comment by lode** [comment]
I would say this is not something that should be altered in Zed's implementation, but instead changed in those `.editorconfig` files. The specification (https://github.com/editorconfig/editorconfig/wiki/EditorConfig-Properties#max_line_length) states `max_line_length` should _force hard_ wrapping.

Though I also see lots of long discussions over at editorconfig (https://github.com/editorconfig/editorconfig/issues/387) about this. But my take would be to follow the current specification (and the good use cases for this property, e.g. markdown) and use other (custom) rules if one wants other features.

**Comment by teohhanhui** [comment]
The problem then is that Zed's EditorConfig implementation incorrectly uses soft wrapping instead of the expected hard wrapping...

**Comment by SomeoneToIgnore** [comment]
Ok, no more wraps for us then: https://github.com/zed-industries/zed/pull/20198
Zed does not have any "hard wrap" settings at all, ergo we'll ignore this now.

**Pull request merged**